File: Rost/src/cube.rs
#![allow(dead_code)]

use std::cmp;
use colored::Colorize;
use crate::{enums::{CubeColor, CubeMove, u8_to_CubeColor}, side::{SOLVED_SIDES, Side}, side};

const SIDE_COUNT: i32 = 6;
const LEFT: i32  = 0;
const RIGHT: i32  = 1;
const BOTTOM: i32  = 2;
const TOP: i32  = 3;
const FRONT: i32  = 4;
const BACK: i32  = 5;

//Adj stripes of adj sides (Clockwise order)
struct SideStripe(i32, i32);
const ADJ_STRIPES: [[SideStripe; 4]; 6] = [
    [
        SideStripe(BOTTOM,    1),
        SideStripe(BACK,      3),
        SideStripe(TOP,       3),
        SideStripe(FRONT,     3)
    ],
    [
        SideStripe(BOTTOM,    3),
        SideStripe(FRONT,     1),
        SideStripe(TOP,       1),
        SideStripe(BACK,      1)
    ],
    [
        SideStripe(LEFT,      3),
        SideStripe(FRONT,     2),
        SideStripe(RIGHT,     1),
        SideStripe(BACK,      0)
    ],
    [
        SideStripe(LEFT,      1),
        SideStripe(BACK,      2),
        SideStripe(RIGHT,     3),
        SideStripe(FRONT,     0)
    ],
    [
        SideStripe(LEFT,      2),
        SideStripe(TOP,       2),
        SideStripe(RIGHT,     2),
        SideStripe(BOTTOM,    2)
    ],
    [
        SideStripe(LEFT,      0),
        SideStripe(BOTTOM,    0),
        SideStripe(RIGHT,     0),
        SideStripe(TOP,       0)
    ],
];

struct SideSquare(i32, i32, i32, i32);

const ORANGE_BLOCKS: [SideSquare; 4] = [
    SideSquare(TOP,       3, BACK,    2),
    SideSquare(TOP,       2, FRONT,   3),
    SideSquare(BOTTOM,    1, FRONT,   2),
    SideSquare(BOTTOM,    0, BACK,    3)
];
const RED_BLOCKS: [SideSquare; 4] = [
    SideSquare(BOTTOM,    3, BACK,    0),
    SideSquare(BOTTOM,    2, FRONT,   1),
    SideSquare(TOP,       1, FRONT,   0),
    SideSquare(TOP,       0, BACK,    1)
];

const ADJ_EDGES: [[i32; 6]; 6] = [
    //Left
    [ -1, -1, 7, 3, 5, 1 ],
    //Right
    [ -1, -1, 3, 7, 5, 1 ],

    //Bottom
    [ 3, 7, -1, -1, 5, 1 ],
    //Top
    [ 7, 3, -1, -1, 5, 1 ],

    //Front
    [ 7, 3, 5, 1, -1, -1 ],
    //Back
    [ 7, 3, 1, 5, -1, -1 ],
];

const ADJ_CORNERS: [[i32; 12]; 6] = [
        //Left						removed cuz redundancy
        [ -1, -1, -1, -1, 6, 0, 4, 2, -2, -2, -2, -2 ],
        //Right
        [ -1, -1, -1, -1, 4, 2, 6, 0, -2, -2, -2, -2 ],

        //Bottom
        [ 4, 2, 6, 0, -1, -1, -1, -1, -2, -2, -2, -2 ],
        //Top
        [ 6, 0, 4, 2, -1, -1, -1, -1, -2, -2, -2, -2 ],

        //Front
        [ 6, 0, 4, 2, -2, -2, -2, -2, -1, -1, -1, -1],
        //Back
        [ 0, 6, 2, 4, -2, -2, -2, -2, -1, -1, -1, -1],
];

#[derive(Eq, PartialEq, Hash, Clone)]
pub struct Cube{
    sides: [Side; 6],
}

impl Cube {
    pub fn new() -> Self {
        Cube{sides: [
                SOLVED_SIDES[0],
                SOLVED_SIDES[1],
                SOLVED_SIDES[2],
                SOLVED_SIDES[3],
                SOLVED_SIDES[4],
                SOLVED_SIDES[5],
            ]}
    }

    pub fn reset(&mut self){
        for i in 0..6 {
            self.sides[i] = SOLVED_SIDES[i];
        }
    }

    pub fn make_move(&mut self, m: CubeMove){
        let side: i32 = (m as i32) / 3;

        let move_type: i32 = m as i32;
        self.sides[side as usize].rotate((move_type % 3) + 1);

        
        let stripes = &ADJ_STRIPES[side as usize];
        match move_type  {
            0 => {
                let buffer = self.sides[stripes[0].0 as usize].get_stripe(stripes[0].1);

                self.sides[stripes[0].0 as usize].set_stripe(stripes[0].1,
                    self.sides[stripes[3].0 as usize].get_stripe(stripes[3].1));

                self.sides[stripes[3].0 as usize].set_stripe(stripes[3].1,
                    self.sides[stripes[2].0 as usize].get_stripe(stripes[2].1));

                self.sides[stripes[2].0 as usize].set_stripe(stripes[2].1,
                    self.sides[stripes[1].0 as usize].get_stripe(stripes[1].1));

                self.sides[stripes[1].0 as usize].set_stripe(stripes[1].1, buffer);
            },
            1 => {
                let mut buffer = self.sides[stripes[0].0 as usize].get_stripe(stripes[0].1);
                self.sides[stripes[0].0 as usize].set_stripe(stripes[0].1,
                    self.sides[stripes[2].0 as usize].get_stripe(stripes[2].1));
                self.sides[stripes[2].0 as usize].set_stripe(stripes[2].1, buffer);
    
                buffer = self.sides[stripes[1].0 as usize].get_stripe(stripes[1].1);
                self.sides[stripes[1].0 as usize].set_stripe(stripes[1].1,
                    self.sides[stripes[3].0 as usize].get_stripe(stripes[3].1));
                self.sides[stripes[3].0 as usize].set_stripe(stripes[3].1, buffer);
            },
            3 => {
                let buffer = self.sides[stripes[0].0 as usize].get_stripe(stripes[0].1);

                self.sides[stripes[0].0 as usize].set_stripe(stripes[0].1,
                    self.sides[stripes[1].0 as usize].get_stripe(stripes[1].1));

                self.sides[stripes[1].0 as usize].set_stripe(stripes[1].1,
                    self.sides[stripes[2].0 as usize].get_stripe(stripes[2].1));

                self.sides[stripes[2].0 as usize].set_stripe(stripes[2].1,
                    self.sides[stripes[3].0 as usize].get_stripe(stripes[3].1));

                self.sides[stripes[3].0 as usize].set_stripe(stripes[3].1, buffer);
            },

            _ => {},
        }
    }

    pub fn is_solved(&self) -> bool {
        for i in 0..6 {
            if self.sides[i] != SOLVED_SIDES[i] {
                return  false;
            }
        }

        true
    }

    pub fn get_edge_color(&self, side1: i32, side2: i32) -> CubeColor {
        self.sides[side1 as usize].get_color(ADJ_EDGES[side1 as usize][side2 as usize])
    }
    pub fn set_edge_color(&mut self, side1: i32, side2: i32, color: CubeColor){
        self.sides[side1 as usize].set_color(ADJ_EDGES[side1 as usize][side2 as usize], color);
    }

    pub fn get_corner_color(&self, side1: i32, side2: i32, side3: i32) -> CubeColor {
        let lower: i32 = cmp::min(side2, side3);
        let higher: i32 = cmp::max(side2, side3);

        self.sides[side1 as usize].get_color(ADJ_CORNERS[side1 as usize][(lower * 2 + (higher & 1)) as usize])
    }

    pub fn set_corner_color(&mut self, side1: i32, side2: i32, side3: i32, color: CubeColor) {
        let lower: i32 = cmp::min(side2, side3);
        let higher: i32 = cmp::max(side2, side3);

        self.sides[side1 as usize].set_color(ADJ_CORNERS[side1 as usize][(lower * 2 + (higher & 1)) as usize], color);
    }

    pub fn edge_is_oriented(&self, side1: i32, side2: i32) -> bool {
        debug_assert!(side1 < side2);

        let c1 = self.get_edge_color(side1, side2) as u8;
        let c2 = self.get_edge_color(side2, side1) as u8;

        //if contains white or yellow
        if c1 / 2 == 1 || c2 / 2 == 1 {
            if side1 / 2 == 0 {
                c2 / 2 == 1
            }
            else {
                c1 / 2 == 1
            }
        }
        else {
            if side1 / 2 == 0 {
                c2 / 2 == 2
            }
            else {
                c1 / 2 == 2
            }
        }
    }

    pub fn corner_orientation(&self, side1: i32, side2: i32, side3: i32) -> i32 {
        debug_assert!(side1 < side2 && side2 < side3);

        if self.get_corner_color(side1, side2, side3) as i32 / 2 == 0 {
            return 0;
        }

        if self.get_corner_color(side2, side1, side3) as i32 / 2 == 0 {
            return 1;
        }

        2
    }

    pub fn get_side_view(&self) -> String{
        let mut s = " ".repeat(5) + &"-".repeat(7) + "\n";

        for y in 0..3 {
            s .push_str(&" ".repeat(5));
            s .push('|');

            for x in 0..3 {
                if x == 1 && y == 1 {
                    s.push(CubeColor::Blue.to_string());
                }
                else {
                    s.push(self.sides[BACK as usize].get_color_2d(x, y).to_string());
                }

                s.push(if x == 2 {'|'} else {' '});
            }
            s.push('\n');
        }


        s.push_str(&"-".repeat(24));
        s.push('\n');
        let inidces: [u8; 4] = [0, 3, 1, 2];

        for y in 0..3 {
            for index in inidces{
                for x in 0..3 {
                    if x == 1 && y == 1 {
                        s.push(u8_to_CubeColor(index).to_string())
                    }
                    else {
                        s.push(self.sides[index as usize].get_color_2d(x, y).to_string());
                    }

                    s.push_str(if x == 2 {"|"} else {" "});
                }
            }

            s.push('\n');
        }

        s.push_str(&"-".repeat(24));
        s.push('\n');
        for y in 0..3 {
            s .push_str(&" ".repeat(5));
            s .push_str("|");

            for x in 0..3 {
                if x == 1 && y == 1 {
                    s.push(CubeColor::Green.to_string());
                }
                else {
                    s.push(self.sides[FRONT as usize].get_color_2d(x, y).to_string());
                }

                s.push(if x == 2 {'|'} else {' '});
            }
            s.push('\n');
        }

        s
    }

    pub fn print_side_view(&self){
        let s = self.get_side_view();

        for c in s.chars() {
            match c {
                'O' => print!("{}", "O".truecolor(255, 127, 0)), 
                'R' => print!("{}", "R".red()), 
                'Y' => print!("{}", "Y".yellow()), 
                'W' => print!("{}", "W".white()), 
                'G' => print!("{}", "G".green()), 
                'B' => print!("{}", "B".blue()), 

                _ => print!("{}", c.to_string().bright_black()) 
            }
        }
    }

    pub fn print_raw(&self){
        for i in 0..6 {
            self.sides[i].print_raw();
            println!();
        }
    }
}

File: CSharp/CubeAD/Cube.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.Linq;

namespace CubeAD
{
	/// <summary>
	/// A class representing a 3x3x3 rubiks cube storing the color of each sticker
	/// </summary>
	public class Cube
	{
		const int SIDE_COUNT = 6;
		const int LEFT = 0;
		const int RIGHT = 1;
		const int BOTTOM = 2;
		const int TOP = 3;
		const int FRONT = 4;
		const int BACK = 5;

		//The side values of a solved Cube
		private static readonly Side[] SolvedColors =
{
			new Side((CubeColor)0),
			new Side((CubeColor)1),
			new Side((CubeColor)2),
			new Side((CubeColor)3),
			new Side((CubeColor)4),
			new Side((CubeColor)5)
		};

		//Adjacent stripes of adj sides (Clockwise order)
		private static readonly (int Side, int Stripe)[][] AdjStripes = new (int Side, int Stripe)[][]
		{
			//Left
			new (int Side, int Stripe)[]
			{
				(BOTTOM,    1),
				(BACK,      3),
				(TOP,       3),
				(FRONT,     3)
			},

			//Right
			new (int Side, int Stripe)[]
			{
				(BOTTOM,    3),
				(FRONT,     1),
				(TOP,       1),
				(BACK,      1)
			},

			//Bottom
			new (int Side, int Stripe)[]
			{
				(LEFT,      3),
				(FRONT,     2),
				(RIGHT,     1),
				(BACK,      0)
			},

			//Top
			new (int Side, int Stripe)[]
			{
				(LEFT,      1),
				(BACK,      2),
				(RIGHT,     3),
				(FRONT,     0)
			},

			//Front
			new (int Side, int Stripe)[]
			{
				(LEFT,      2),
				(TOP,       2),
				(RIGHT,     2),
				(BOTTOM,    2)
			},

			//Back
			new (int Side, int Stripe)[]
			{
				(LEFT,      0),
				(BOTTOM,    0),
				(RIGHT,     0),
				(TOP,       0)
			}
		};

		//Squares that form a block (Clockwise after squares of main side)
		private static (int side1, int square1, int side2, int square2)[] OrangeBlocks =
		{
			(TOP,       3, BACK,    2),
			(TOP,       2, FRONT,   3),
			(BOTTOM,    1, FRONT,   2),
			(BOTTOM,    0, BACK,    3)
		};
		private static (int side1, int square1, int side2, int square2)[] RedBlocks =
		{
			(BOTTOM,    3, BACK,    0),
			(BOTTOM,    2, FRONT,   1),
			(TOP,       1, FRONT,   0),
			(TOP,       0, BACK,    1)
		};

		//SIDE_COUNT Sides of the cube
		Side[] Sides = new Side[SIDE_COUNT];

		/// <summary>
		/// Creates a solved cube
		/// </summary>
		public Cube() 
		{
			Reset();
		}
		public void Reset()
		{
			for (int i = 0; i < SIDE_COUNT; i++)
			{
				Sides[i] = new Side((CubeColor)i);
			}
		}

		/// <summary>
		/// Creates a copy of a cube
		/// </summary>
		public Cube(Cube old) 
		{
			for (int i = 0; i < SIDE_COUNT; i++)
			{
				Sides[i] = old.Sides[i];
			}
		}

		/// <summary>
		/// Creates copy of a <see cref="Cube"/> and makes a <see cref="CubeMove"/> afterwards
		/// </summary>
		public Cube(Cube old, CubeMove m)
		{
			for (int i = 0; i < SIDE_COUNT; i++)
			{
				Sides[i] = old.Sides[i];
			}

			MakeMove(m);
		}
		
		public void CopyValuesFrom(Cube c)
		{
			for (int i = 0; i < SIDE_COUNT; i++)
			{
				Sides[i] = c.Sides[i];
			}
		}

		public void ApplyMoveSequenz(MoveSequenz s)
		{
			for (int i = 0; i < s.Moves.Count; i++)
				MakeMove(s.Moves[i]);
		}

		public void MakeMove(CubeMove m)
		{
			int side = (int)m / 3;

			//Rotate affected side
			Sides[side].Rotate(((int)m % 3) + 1);

			uint buffer;
			int moveType = (int)m % 3;
			(int Side, int Stripe)[] stripes = AdjStripes[side];

			//Swap stickers from adjacent sides
			switch (moveType)
			{
				case 0:
					buffer = Sides[stripes[0].Side].GetStripe(stripes[0].Stripe);

					Sides[stripes[0].Side].SetStripe(stripes[0].Stripe,
						Sides[stripes[3].Side].GetStripe(stripes[3].Stripe));

					Sides[stripes[3].Side].SetStripe(stripes[3].Stripe,
						Sides[stripes[2].Side].GetStripe(stripes[2].Stripe));

					Sides[stripes[2].Side].SetStripe(stripes[2].Stripe,
						Sides[stripes[1].Side].GetStripe(stripes[1].Stripe));

					Sides[stripes[1].Side].SetStripe(stripes[1].Stripe, buffer);
					break;

				case 1:
					buffer = Sides[stripes[0].Side].GetStripe(stripes[0].Stripe);
					Sides[stripes[0].Side].SetStripe(stripes[0].Stripe,
						Sides[stripes[2].Side].GetStripe(stripes[2].Stripe));
					Sides[stripes[2].Side].SetStripe(stripes[2].Stripe, buffer);

					buffer = Sides[stripes[1].Side].GetStripe(stripes[1].Stripe);
					Sides[stripes[1].Side].SetStripe(stripes[1].Stripe,
						Sides[stripes[3].Side].GetStripe(stripes[3].Stripe));
					Sides[stripes[3].Side].SetStripe(stripes[3].Stripe, buffer);
					break;

				case 2:
					buffer = Sides[stripes[0].Side].GetStripe(stripes[0].Stripe);

					Sides[stripes[0].Side].SetStripe(stripes[0].Stripe,
						Sides[stripes[1].Side].GetStripe(stripes[1].Stripe));

					Sides[stripes[1].Side].SetStripe(stripes[1].Stripe,
						Sides[stripes[2].Side].GetStripe(stripes[2].Stripe));

					Sides[stripes[2].Side].SetStripe(stripes[2].Stripe,
						Sides[stripes[3].Side].GetStripe(stripes[3].Stripe));

					Sides[stripes[3].Side].SetStripe(stripes[3].Stripe, buffer);
					break;
			}
		}

		public bool IsSolved
		{
			get
			{
				for (int i = 0; i < SIDE_COUNT; i++)
					if (Sides[i] != SolvedColors[i]) return false;

				return true;
			}
		}

		//The index of the sticker on the first side which is adjacent to the second side 
		public static int[,] AdjEdges = new int[SIDE_COUNT, SIDE_COUNT]
		{
			//Left
			{ -1, -1, 7, 3, 5, 1 },
			//Right
			{ -1, -1, 3, 7, 5, 1 },

			//Bottom
			{ 3, 7, -1, -1, 5, 1 },
			//Top
			{ 7, 3, -1, -1, 5, 1 },

			//Front
			{ 7, 3, 5, 1, -1, -1 },
			//Back
			{ 7, 3, 1, 5, -1, -1 },
		};

		/// <returns>The color of the sticker on <paramref name="side1"/> adjacent to <paramref name="side2"/> </returns>
		public CubeColor GetEdgeColor(int side1, int side2)
		{
			return (CubeColor)Sides[side1][AdjEdges[side1, side2]];
		}
		/// <summary>
		/// Set the color of the sticker on <paramref name="side1"/> adjacent to <paramref name="side2"/> to <paramref name="color"/> 
		/// </summary>
		public void SetEdgeColor(int side1, int side2, int color)
		{
			Sides[side1][AdjEdges[side1, side2]] = (uint)color;
		}

		//The index of the sticker on the first side which is adjacent to the second and third side
		//Only the parity of the last side is considered
		public static int[,] AdjCorners = new int[SIDE_COUNT, SIDE_COUNT * 2]
		{
			//Left						removed because of redundancy
			{ -1, -1, -1, -1, 6, 0, 4, 2, -2, -2, -2, -2 },
			//Right
			{ -1, -1, -1, -1, 4, 2, 6, 0, -2, -2, -2, -2 },

			//Bottom
			{ 4, 2, 6, 0, -1, -1, -1, -1, -2, -2, -2, -2 },
			//Top
			{ 6, 0, 4, 2, -1, -1, -1, -1, -2, -2, -2, -2 },

			//Front
			{ 6, 0, 4, 2, -2, -2, -2, -2, -1, -1, -1, -1},
			//Back
			{ 0, 6, 2, 4, -2, -2, -2, -2, -1, -1, -1, -1},
		};

		/// <returns>The color of the sticker on <paramref name="side1"/> adjacent to <paramref name="side2"/> and <paramref name="side3"/></returns>
		public CubeColor GetCornerColor(int side1, int side2, int side3)
		{
			int lower = Math.Min(side2, side3);
			int higher = Math.Max(side2, side3);

			return (CubeColor)Sides[side1][AdjCorners[side1, lower * 2 + (higher & 1)]];
		}

		/// <summary>
		/// Set the color of the sticker on <paramref name="side1"/> adjacent to <paramref name="side2"/> and <paramref name="side3"/> to <paramref name="color"/> 
		/// </summary>
		public void SetCornerColor(int side1, int side2, int side3, int color)
		{
			int lower = Math.Min(side2, side3);
			int higher = Math.Max(side2, side3);

			Sides[side1][AdjCorners[side1, lower * 2 + (higher & 1)]] = (uint)color;
		}

		/// <summary>
		///  Checks whether this <see cref="Cube"/> is self symmetric
		/// </summary>
		/// <param name="se">Symmetry to check for</param>
		public bool HasSymmetry(SymmetryElement se)
		{
			//Check edges
			for (int x = 0; x < SIDE_COUNT; x++)
			{
				int nextX = se.TransformColor(x);
				for (int y = 0; y < SIDE_COUNT; y++)
				{
					if (x / 2 == y / 2) continue;

					int nextY = se.TransformColor(y);
					
					//Check whether the sticker has the correct color after the transformation and the color change
					if (se.TransformColor(GetEdgeColor(x, y)) != GetEdgeColor(nextX, nextY))
						return false;
				}
			}

			//Check corners
			for (int x = 0; x < SIDE_COUNT; x++)
			{
				int nextX = se.TransformColor(x);
				for (int y = 0; y < SIDE_COUNT; y++)
				{
					if (x / 2 == y / 2) continue;
					int nextY = se.TransformColor(y);

					for (int z = 0; z < SIDE_COUNT; z++)
					{
						if (x / 2 == z / 2 || y / 2 == z / 2) continue;
						int nextZ = se.TransformColor(z);

						//Check whether the sticker has the correct color after the transformation and the color change
						if (se.TransformColor(GetCornerColor(x, y, z)) != GetCornerColor(nextX, nextY, nextZ))
							return false;
					}

				}
			}

			return true;
		}

		/// <returns>A bit vector containing a 1 for each <see cref="SymmetryElement"/> this <see cref="Cube"/> has</returns>
		public BitMap64 GetSymmetrySet()
		{
			BitMap64 ret = new BitMap64();

			for(int i = 0; i < SymmetryElement.ORDER; i++) {
				ret[i] = HasSymmetry(SymmetryElement.Elements[i]);
			}

			return ret;
		}

		/// <returns>Whether this <see cref="Cube"/> is equal to <paramref name="other"/> with any symmetry </returns>
		public bool IsEqualWithSymmetry(Cube other)
		{
			for (int i = 0; i < SymmetryElement.ORDER; i++)
			{
				if (IsEqualWithSymmetry(other, SymmetryElement.Elements[i]))
					return true;
			}

			return false;
		}


		/// <summary>
		/// Computes whether this <see cref="Cube"/> is symmetric to <paramref name="other"/> over <paramref name="se"/>
		/// </summary>
		public bool IsEqualWithSymmetry(Cube other, SymmetryElement se)
		{
			//Check edges
			for (int x = 0; x < 6; x++)
			{
				int nextX = se.TransformColor(x);
				for (int y = 0; y < 6; y++)
				{
					if (x / 2 == y / 2) continue;

					int nextY = se.TransformColor(y);

					//Check whether the sticker has the correct color after the transformation and the color change
					if (se.TransformColor(GetEdgeColor(x, y)) != other.GetEdgeColor(nextX, nextY))
						return false;
				}
			}

			//Check corners
			for (int x = 0; x < 6; x++)
			{
				int nextX = se.TransformColor(x);
				for (int y = 0; y < 6; y++)
				{
					if (x / 2 == y / 2) continue;
					int nextY = se.TransformColor(y);

					for (int z = 0; z < 6; z++)
					{
						if (x / 2 == z / 2 || y / 2 == z / 2) continue;
						int nextZ = se.TransformColor(z);

						//Check whether the sticker has the correct color after the transformation and the color change
						if (se.TransformColor(GetCornerColor(x, y, z)) != other.GetCornerColor(nextX, nextY, nextZ))
							return false;
					}

				}
			}

			return true;
		}

		/// <summary>
		/// Copies over <paramref name="se"/> transformed stickers to <paramref name="cube"/>
		/// </summary>
		public void InsertSymmetryTransformation(SymmetryElement se, Cube cube)
		{
			//Check edges
			for (int x = 0; x < 6; x++)
			{
				int nextX = se.TransformColor(x);
				for (int y = 0; y < 6; y++)
				{
					if (x / 2 == y / 2) continue;

					int nextY = se.TransformColor(y);

					cube.SetEdgeColor(nextX, nextY, (int)se.TransformColor(GetEdgeColor(x, y)));
				}
			}

			//Check corners
			for (int x = 0; x < 6; x++)
			{
				int nextX = se.TransformColor(x);
				for (int y = 0; y < 6; y++)
				{
					if (x / 2 == y / 2) continue;
					int nextY = se.TransformColor(y);

					for (int z = 0; z < 6; z++)
					{
						if (x / 2 == z / 2 || y / 2 == z / 2) continue;
						int nextZ = se.TransformColor(z);

						cube.SetCornerColor(nextX, nextY, nextZ, (int)se.TransformColor(GetCornerColor(x, y, z)));
					}
				}
			}
		}

		/// <returns> A cube net like representation</returns>
		public string GetSideView()
		{
			string s = new string(' ', 3 * 2 - 1) + new string('-', 3 * 2 + 1) + "\n";
			for (int y = 0; y < 3; y++)
			{
				s += new string(' ', 3 * 2 - 1) + "|";
				for (int x = 0; x < 3; x++)
				{
					if (x == 1 && y == 1)
						s += CubeColor.Blue.ToString()[0];
					else
						s += Sides[BACK][x, y].ToString()[0];

					s += x == 2 ? "|" : " ";
				}

				s += "\n";
			}

			s += new string('-', (3 * 2) * 4) + "\n";
			int[] Indices = { 0, 3, 1, 2 };
			for (int y = 0; y < 3; y++)
			{
				for (int i = 0; i < Indices.Length; i++)
				{
					for (int x = 0; x < 3; x++)
					{
						if (x == 1 && y == 1)
							s += ((CubeColor)Indices[i]).ToString()[0];
						else
							s += Sides[Indices[i]][x, y].ToString()[0];

						s += x == 2 ? "|" : " ";
					}


				}

				s += "\n";
			}
			s += new string('-', (3 * 2) * 4) + "\n";

			for (int y = 0; y < 3; y++)
			{
				s += new string(' ', 3 * 2 - 1) + "|";
				for (int x = 0; x < 3; x++)
				{
					if (x == 1 && y == 1)
						s += CubeColor.Green.ToString()[0];
					else
						s += Sides[FRONT][x, y].ToString()[0];

					s += x == 2 ? "|" : " ";
				}

				s += "\n";
			}
			s += new string(' ', 3 * 2 - 1) + new string('-', 3 * 2 + 1);

			return s + "\n";
		}

		/// <summary>
		/// Prints a colored side view of this <see cref="Cube"/> to the console
		/// </summary>
		public void PrintSideView()
		{
			string s = GetSideView();
			for (int i = 0; i < s.Length; i++)
			{
				Color color;
				switch (s[i])
				{
					case 'W': color = Color.White; break;
					case 'R': color = Color.Red; break;
					case 'O': color = Color.Orange; break;
					case 'B': color = Color.Blue; break;
					case 'G': color = Color.Green; break;
					case 'Y': color = Color.Yellow; break;
					default: color = Color.Gray; break;
				}

				Colorful.Console.Write(s[i], color);
			}

			Console.ForegroundColor = ConsoleColor.White;
			Colorful.Console.Write("\n", Color.White);
		}

		/// <returns>The orientaion of the edge at <paramref name="side1"/>/<paramref name="side2"/></returns>
		public bool EdgeIsOriented(int side1, int side2)
		{
			if (side2 < side1)
				(side1, side2) = (side2, side1);

			int c1 = (int)GetEdgeColor(side1, side2);
			int c2 = (int)GetEdgeColor(side2, side1);


			//if contains white or yellow
			if (c1 / 2 == 1 || c2 / 2 == 1)
			{
				if (side1 / 2 == 0)
					return c2 / 2 == 1;
				else
					return c1 / 2 == 1;
			}
			else
			{
				if (side1 / 2 == 0)
					return c2 / 2 == 2;
				else
					return c1 / 2 == 2;
			}
		}

		/// <returns>The orientaion of the corner at <paramref name="side1"/>/<paramref name="side2"/>/<paramref name="side3"/></returns>
		public int CornerOrientaion(int side1, int side2, int side3)
		{
			if (!(side1 < side2 && side2 < side3))
				throw new ArgumentException("Sides need to be in ascending order");

			if ((int)GetCornerColor(side1, side2, side3) / 2 == 0)
				return 0;

			if ((int)GetCornerColor(side2, side1, side3) / 2 == 0)
				return 1;

			return 2;
		}

		public static bool operator ==(Cube a, Cube b)
		{
			for (int i = 0; i < 6; i++)
				if (a.Sides[i] != b.Sides[i]) return false;

			return true;
		}
		public static bool operator !=(Cube a, Cube b)
		{
			return !(a == b);
		}
		public override bool Equals(object obj)
		{
			Cube c = (Cube)obj;

			for (int i = 0; i < 6; i++)
			{
				if (Sides[i] != c.Sides[i]) return false;
			}
			return true;
		}

		public override int GetHashCode()
		{
			return HashCode.Combine(Sides[0], Sides[1], Sides[2], Sides[3], Sides[4], Sides[5]);
		}
	}
}
